modules/dag-l0/src/main/scala/org/tessellation/dag/l0/domain/snapshot/programs/GlobalSnapshotEventCutter.scala

```diff
@@ -0,0 +1,144 @@
+        def hash: Hash = event.hash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent = event.signed.asRight[StateChannelEvent]
+
+        def isMyChild(that: PeerIdHashedEvent): Boolean =
+          that.event.signed.value.parent.contains_(event.ownReference)
+      }
+
+      class OrderedEvents(sortedStateEvents: List[PeerIdStateChannelEvent], hashedEvents: List[PeerIdHashedEvent]) {
+        def next: Option[(PeerIdEvent, OrderedEvents)] = {
+          val (parents, nonParents) =
+            hashedEvents.foldLeft((List.empty[PeerIdHashedEvent], List.empty[PeerIdHashedEvent])) {
+              case ((parents, nonParents), p) if hashedEvents.exists(p.isMyChild) => (p :: parents, nonParents)
+              case ((parents, nonParents), p)                                     => (parents, p :: nonParents)
+            }
+
+          (sortedStateEvents, nonParents.sorted(ordering)) match {
+            case (Nil, Nil) =>
+              none
+            case (h :: t, Nil) =>
+              (h, new OrderedEvents(t, Nil)).some
+            case (Nil, h :: t) =>
+              (h, new OrderedEvents(Nil, t ++ parents)).some
+            case (h1 :: t1, h2 :: t2) =>
+              if (h1.fee < h2.fee || (h1.fee == h2.fee && sortedStateEvents.length >= hashedEvents.length))
+                (h1, new OrderedEvents(t1, hashedEvents)).some
+              else
+                (h2, new OrderedEvents(sortedStateEvents, t2 ++ parents)).some
+          }
+        }
+      }
+
+      object OrderedEvents {
+        // NOTE: Order for removal: NOT parent → Fee → Split by event type → Hash
+        def apply(events: GlobalPeerIdEventMap): F[OrderedEvents] =
+          unpackGlobalPeerIdEvents(events).map {
+            case (stateChannelEvents, hashedEvents) =>
+              new OrderedEvents(stateChannelEvents.sorted(ordering), hashedEvents)
+          }
+
+        def unpackGlobalPeerIdEvents(events: GlobalPeerIdEventMap): F[(List[PeerIdStateChannelEvent], List[PeerIdHashedEvent])] =
+          events.view
+            .mapValues(es => unpackEither(es.map(_._2)))
+            .toList
+            .traverse {
+              case (peerId, (stateChannelEvents, dagEvents)) =>
+                dagEvents.traverse(_.toHashed[F]).map(hashedEvents => peerId -> (stateChannelEvents, hashedEvents))
+            }
+            .map { es =>
+              val z = (List.empty[PeerIdStateChannelEvent], List.empty[PeerIdHashedEvent])
+              es.foldLeft(z) {
+                case ((accLeft, accRight), (peerId, (stateChannelEvents, hashedEvents))) =>
+                  val left = stateChannelEvents.map(PeerIdStateChannelEvent(peerId, _))
+                  val right = hashedEvents.map(PeerIdHashedEvent(peerId, _))
+                  (left ++ accLeft, right ++ accRight)
+              }
+            }
+
+        def unpackEither[A, B](es: List[Either[A, B]]): (List[A], List[B]) =
+          es.foldLeft((List.empty[A], List.empty[B])) {
```

Comment:
I think you can use `.partitionMap(identity)` here

Reply:
@TheMMaciek thank you, I was looking for a better way to do that.